std-bits/table/src/main/java/org/enso/table/excel/xssfreader/XSSFReaderSheet.java

```diff
@@ -0,0 +1,163 @@
+import java.util.TreeMap;
+import javax.xml.parsers.ParserConfigurationException;
+import org.apache.poi.openxml4j.exceptions.InvalidFormatException;
+import org.apache.poi.ss.usermodel.Sheet;
+import org.apache.poi.util.XMLHelper;
+import org.enso.table.excel.ExcelRow;
+import org.enso.table.excel.ExcelSheet;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+
+public class XSSFReaderSheet implements ExcelSheet {
+  private final int sheetIdx;
+  private final String sheetName;
+  private final String relId;
+  private final XSSFReaderWorkbook parent;
+
+  private boolean readSheetData = false;
+  private String dimensions;
+  private int firstRow;
+  private int lastRow;
+  private Map<Integer, SortedMap<Short, XSSFReaderSheetXMLHandler.CellValue>> rowData;
+
+  public XSSFReaderSheet(int sheetIdx, String sheetName, String relId, XSSFReaderWorkbook parent) {
+    this.sheetIdx = sheetIdx;
+    this.sheetName = sheetName;
+    this.relId = relId;
+    this.parent = parent;
+  }
+
+  private synchronized void readSheetData() {
+    if (readSheetData) {
+      return;
+    }
+
+    try {
+      var strings = parent.getSharedStrings();
+      var styles = parent.getStyles();
+      var handler =
+          new XSSFReaderSheetXMLHandler(styles, strings) {
+            @Override
+            protected void onDimensions(String dimension) {
+              handleOnDimensions(dimension);
+            }
+
+            @Override
+            protected void onStartRow(int rowNum) {
+              handleOnStartRow(rowNum);
+            }
+
+            @Override
+            protected void onCell(int rowNumber, short columnNumber, String ref, CellValue value) {
+              handleOnCell(rowNumber, columnNumber, value);
+            }
+          };
+
+      var xmlReader = XMLHelper.newXMLReader();
+      xmlReader.setContentHandler(handler);
+
+      rowData = new HashMap<>();
+
+      try {
+        parent.withReader(
+            reader -> {
+              try {
+                var sheet = reader.getSheet(relId);
+                xmlReader.parse(new InputSource(sheet));
+              } catch (SAXException | InvalidFormatException | IOException e) {
+                throw new RuntimeException(e);
+              }
+            });
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+
+      readSheetData = true;
+    } catch (SAXException | ParserConfigurationException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public int getSheetIndex() {
+    return sheetIdx;
+  }
+
+  @Override
+  public String getName() {
+    return sheetName;
+  }
+
+  public String getDimensions() {
+    if (!readSheetData) {
+      readSheetData();
+    }
+    return dimensions;
+  }
+
+  @Override
+  public int getFirstRow() {
+    if (!readSheetData) {
+      readSheetData();
+    }
+    return firstRow;
+  }
+
+  @Override
+  public int getLastRow() {
+    if (!readSheetData) {
```

Comment:
given that we repeat these 3 lines of code in so many places, I'd usually create a small helper `ensureWasRead()` that contains the if, so the duplicated part is a simple method call - easier to read, easier to change if need to refactor